tls/s2n_resume.c

```diff
@@ -151,32 +151,90 @@ int s2n_client_serialize_resumption_state(struct s2n_connection *conn, struct s2
     return 0;
 }
 
-static int s2n_client_deserialize_session_state(struct s2n_connection *conn, struct s2n_stuffer *from)
-{
-    if (s2n_stuffer_data_available(from) < S2N_STATE_SIZE_IN_BYTES) {
-        POSIX_BAIL(S2N_ERR_INVALID_SERIALIZED_SESSION_STATE);
-    }
-
-    uint8_t format;
-    uint64_t then;
+static S2N_RESULT s2n_tls12_client_deserialize_session_state(struct s2n_connection *conn, struct s2n_stuffer *from)
+{   
+    RESULT_ENSURE_REF(conn);
+    RESULT_ENSURE_REF(from);
 
-    POSIX_GUARD(s2n_stuffer_read_uint8(from, &format));
-    if (format != S2N_TLS12_SERIALIZED_FORMAT_VERSION) {
-        POSIX_BAIL(S2N_ERR_INVALID_SERIALIZED_SESSION_STATE);
-    }
+    RESULT_ENSURE(s2n_stuffer_data_available(from) == (S2N_STATE_SIZE_IN_BYTES - S2N_STATE_FORMAT_LEN), 
+        S2N_ERR_INVALID_SERIALIZED_SESSION_STATE);
 
-    POSIX_GUARD(s2n_stuffer_read_uint8(from, &conn->actual_protocol_version));
+    RESULT_GUARD_POSIX(s2n_stuffer_read_uint8(from, &conn->actual_protocol_version));
 
     uint8_t *cipher_suite_wire = s2n_stuffer_raw_read(from, S2N_TLS_CIPHER_SUITE_LEN);
-    POSIX_ENSURE_REF(cipher_suite_wire);
-    POSIX_GUARD(s2n_set_cipher_as_client(conn, cipher_suite_wire));
+    RESULT_ENSURE_REF(cipher_suite_wire);
+    RESULT_GUARD_POSIX(s2n_set_cipher_as_client(conn, cipher_suite_wire));
 
-    POSIX_GUARD(s2n_stuffer_read_uint64(from, &then));
+    uint64_t then;
+    RESULT_GUARD_POSIX(s2n_stuffer_read_uint64(from, &then));
```

Comment:
Currently we are not validating this `then` ticket time obtained from the wire, should we evaluate this ticket time obtained in `then` prior to storing the session_data on the connection?

Reply:
This is TLS1.2 behavior, which did does not validate the ticket arrival time. This was a design decision in our TLS1.2 implementation, I think that they wanted the server to be the one to evaluate whether or not a ticket is valid. The TLS1.2 session resumption RFC doesn't have a hard line about this either: https://tools.ietf.org/html/rfc5077#section-5.6. I think we probably shouldn't be changing TLS1.2 behavior although I'd be willing to discuss this.